### PR TITLE
add versioning to shared libs

### PR DESCRIPTION
**Pull request by jonathanspw**
I'm not sure what you have planned for ABI/API compatibility so versioning it to major.minor might be more appropriate than just major.

### CODE REVIEW

**Comment by horenmar** [comment]
ABI stability: NOPE
API stability: semver (modulo bugs and Hyrum's law)

**Comment by dimztimz** [comment]
Becauase there is no ABI stability whatsoever, `SOVERSION` should contain all three parts of the version i.e. `SOVERSION` should be the same with `PROJECT_VERSION`.

**Comment by jonathanspw** [comment]
So even minor versions will potentially change ABI?

That's pretty contradictory to how most ABI versioning works.  Can ABI stability be maintained within X.X releases so that x.x.X can be considered bugfix only, or at least non-breaking?

**Comment by horenmar** [comment]
> So even minor versions will potentially change ABI?

Yes. Catch2 is meant to be used as a static library. As such, it should not be installed globally and ABI stability is a non-concern.

For various reasons, we also support building it as a dynamic library, but there is a number of limitations to it, e.g. symbols are exported only implicitly. Providing ABI stability is a maintenance cost I am not willing to pay.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -301,7 +301,10 @@ if (ANDROID)
     target_link_libraries(Catch2 INTERFACE log)
 endif()
 
-set_target_properties(Catch2 PROPERTIES DEBUG_POSTFIX "d")
+set_target_properties(Catch2 PROPERTIES 
+                          DEBUG_POSTFIX "d"
+                                VERSION ${PROJECT_VERSION}
+                              SOVERSION ${PROJECT_VERSION_MAJOR})
```
Comment (horenmar) [inline src/CMakeLists.txt]:
```suggestion
                              SOVERSION ${PROJECT_VERSION})
```

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -350,6 +353,8 @@ target_link_libraries(Catch2WithMain PUBLIC Catch2)
 set_target_properties(Catch2WithMain
   PROPERTIES
     OUTPUT_NAME "Catch2Main"
+        VERSION ${PROJECT_VERSION}
+      SOVERSION ${PROJECT_VERSION_MAJOR}
```
Comment (horenmar) [inline src/CMakeLists.txt]:
```suggestion
      SOVERSION ${PROJECT_VERSION}
```

**Comment by jonathanspw** [comment]
Thanks for the tweaks and getting this merged!

**Pull request merged**